오픈마켓[STEP1] 백곰, 브래드

bradheo65 opened the pull request:
안녕하세요 콘 @protocorn93 오픈마켓 첫번째 PR 보내드립니다. 🙇🏻‍♂️🙇🏻‍♂️

처음해보는 API 통신이다보니 오래 걸려서 죄송합니다. 잘 부탁드립니다 !

### 기능설명
- UrlCollection: URL주소를 Enum 타입으로 만들어주었습니다
- Network: API 주소에서 가져온 Json 타입의 데이터들을 담을 Model 구조체 입니다
- Page: API 주소에서 가져온 Json 타입의 데이터들을 담을 Model 구조체 입니다
- JsonParser
    - fetch(): 매개변수로 받은 URL 주소를 GET 방식으로 서버와 데이터를 주고받습니다

### 궁금한 점
Q1. 타입 형식에 맞는 JSON데이터 가져오는 법
현재 메서드가 JSON 파싱을 위해 URL 정보 중 key:value에 맞는 타입 값을 지정해줘야 합니다. 하지만 이러한 방법은 메소드 재사용성이 별로 좋지 않다고 생각 하고 있는데 별다른 방법이 떠오르지 않았습니다... 혹시 콘이라면 이러한 상황일 때 어떻게 표현하는게 좋은 지에 대한 생각을 듣고 싶습니다!

### TroubleShooting
1. HTTP method POST 테스트
POST method 를 통해 서버에 테스트 데이터를 API문서를 보면서 필요한 정보에 대해 Json 형식으로 POST를 해주었지만, 계속해서 400 에러가 발생되었습니다. 알고 보니 해당 서버 접근에 필요한 id 와 password가 아직 할당 받지 못해서 발생된 에러였습니다.

bradheo65 committed:
feat: mock 테스트를 위한 모델 타입과 TDD  구현

Baek-Gom-95 committed:
feat : "JsonParser" 파싱 기능 request 구현

bradheo65 committed:
refactor: Step 요구사항 정리

- Step 1-2 진행을 1-1로 진행함에 따른 코드 정리

bradheo65 committed:
feat: API 통신 중 "GET" 메소드 구현

Baek-Gom-95 committed:
refactor : "UrlSession GET"  상품상세 조회  테스트 수정

bradheo65 committed:
docs: README.md, GroundRule, Step01 문서 작성

bradheo65 committed:
Update README.md

protocorn93 commented on OpenMarket/OpenMarket/Enum/UrlCollection.swift:
```diff
@@ -0,0 +1,12 @@
+//  Created by BaekGom, Brad on 2022/07/12.
+//
+
+enum  UrlCollection {
```
Swift에선 URL 모두 대문자로 사용하고 있기에 URLCollection이 좀 더 알맞은 네이밍이 될 것 같아요. 그리고. `enum`과 `UrlCollection` 사이 불필요한 공백이 있네요.

protocorn93 replied:
네임스페이스로 enum 타입을 사용하기로 결정한 이유가 궁금해요~

Baek-Gom-95 replied:
- Url 은 좀 어색했었는데 URL로 네이밍 해주는군요! 공백은 제거했습니다.
- 인스턴스화 할 필요가 없어서 Enum 타입으로 해주었습니다.

protocorn93 commented on OpenMarket/OpenMarket/Json/JsonParser.swift:
```diff
@@ -0,0 +1,41 @@
+
+import Foundation
+
+struct JsonParser {
```
마찬가지로 `JSONParser`가 좀 더 알맞을 것 같아요

protocorn93 commented on OpenMarket/OpenMarket/Json/JsonParser.swift:
```diff
@@ -0,0 +1,41 @@
+            return
+        }
+        
+        var request = URLRequest.init(url: url)
```
init은 제거해줘도 될 것 같아요.

bradheo65 replied:
`URLRequest.(url: url)`로 변경 했습니다!

protocorn93 commented on OpenMarket/OpenMarket/Json/JsonParser.swift:
```diff
@@ -0,0 +1,41 @@
+                result = try JSONDecoder().decode(Page.self, from: data)
+            }
+            catch {
+                print("fail error: \(error)")
```
에러는 바깥에서 처리될 수 있도록 구현해보시는게 좋을 것 같아요.

protocorn93 commented on OpenMarket/OpenMarket/Model/Network.swift:
```diff
@@ -0,0 +1,26 @@
+//  Created by BaekGom, Brad on 2022/07/11.
+//
+
+struct Network: Decodable {
```
이름이 너무 모호해요! 저는 보통 서버 스팩에 네이밍을 맞추는 편이에요. 커뮤니케이션 하기 수월하거든요!  그래서 저 같으면 `ProductListResponse`로 명명할 것 같네요.

bradheo65 replied:
앗 서버 네이밍은 처음이라... 의견 감사합니다! 늘 네이밍에 대해 고민이네요 🤯

protocorn93 commented on OpenMarket/OpenMarket/Model/Page.swift:
```diff
@@ -0,0 +1,43 @@
+    var image: UIImage? {
+        let uiImage = UIImage(named: self.thumbnail)
+        return uiImage
+    }
```
저는 보통 모델도 계층을 나눠요. 

`서버 매핑 모델 - 엔티티 - 뷰 모델 ` 이렇게 모델도 계층을 나눠놓으면 비즈니스 로직이나 서버 응답이 변경되었을 때도 유연하게 대응할 수 있어요

이렇게 나누곤해요. 그래서 실제 뷰에 그려질 때 사용되는 데이터는 뷰모델에서 갖고 있도록 하는 편이에요.

protocorn93 commented on OpenMarket/OpenMarket/Model/Page.swift:
```diff
@@ -0,0 +1,43 @@
+
+import UIKit
+
+struct Page: Decodable {
```
모델명도 `Product`가 더 적합해보이네요.

protocorn93 commented on OpenMarket/OpenMarket/Json/JsonParser.swift:
```diff
@@ -0,0 +1,41 @@
+            guard let json = result else {
+                return
+            }
+            print(json)
```
불필요한 print문은 제거해주시는게 좋을 것 같아요~

bradheo65 replied:
넵 제거 완료 했습니다

protocorn93 commented on OpenMarket/OpenMarket/Json/JsonParser.swift:
```diff
@@ -0,0 +1,41 @@
+        var request = URLRequest.init(url: url)
+        request.httpMethod = "GET"
+        
+        let task = URLSession.shared.dataTask(with: url) { (data, response, error) in
```
여기서 발생할 수 있는 error도 별도로 처리할 필요가 있어보여요

Baek-Gom-95 replied:
escaping 클로저를 사용해 호출한 함수에서 success, fail  로 처리 하도록 변경했습니다

protocorn93 commented on OpenMarket/OpenMarket/Json/JsonParser.swift:
```diff
@@ -0,0 +1,41 @@
+            var result: Page?
+            
+            do {
+                result = try JSONDecoder().decode(Page.self, from: data)
```
fetch 메서드를 사용하는 곳에선 결과를 어떻게 받아볼 수 있을까요?

Baek-Gom-95 replied:
escaping 클로저를 사용해 호출한 함수에서 success 일때 data로 결과를 받아오도록 변경했습니다

protocorn93 replied:
잘해주셨어요! 👍

protocorn93 commented on OpenMarket/JsonTest/JsonTest.swift:
```diff
@@ -0,0 +1,48 @@
+import XCTest
+@testable import OpenMarket
+
+class JsonTest: XCTestCase {
```
현재 테스트 코드를 보면 JSONDecoder, 즉 애플에서 제공하는 기본 컴포넌트를 테스트하고 있어요. 이는 불필요한 테스트 코드에요. 이미 애플에서 수많은 테스트를 거쳐 우리에게 온 컴포넌트이기 때문이에요. 

여기서 두 분이 테스트해야 할 객체는 여러분이 작성한 `JsonParser`에요. 이를 어떻게 테스트 해볼 수 있을까요? 또 실제 네트워크 요청 없이 어떻게 테스트 해볼 수 있을까요?

Baek-Gom-95 replied:
MockURLSession을 만들어서 Unit Test 수정후 테스트 했습니다.

protocorn93 replied:
확인했어요! 잘 작성해주셨네요! 👍

protocorn93 reviewed (changes requested):
두 분 모두 고생하셨어요! 단위 테스트와 확작성에 대해서 조금 고민해볼 필요가 있을 것 같아요! 

단위 테스트에선 실제 네트워크 요청을 하지 않아요. 흔히 말하는 가짜 데이터를 주입해줌으로써 input/output을 검증하는 것이죠. 이 부분은 코멘트 확인 부탁드릴게요. 

그리고 서버가 아니라 로컬에 저장된 제품 리스트를 가져와야 할 땐 어떻게 해야할까요? 또한 로컬 저장소가 CoreData라면 혹은 Realm이라면 어떻게 대응해볼 수 있을까요?? 

## 질문 

> Q1. 타입 형식에 맞는 JSON데이터 가져오는 법

보통 데이터를 가져올 때
1. 외부 저장소(서버)에 접근하는 객체
2. 이를 사용해 화면에서 사용되는 적절한 화면의 데이터로 매핑하는 객체 
이렇게 두 가지 객체를 활용해요. 즉 1번 객체는 여러곳에서 재사용될 수 있고 2번 객체는 특정 화면에서만 사용 혹은 특정 도메인 안에서 재사용될 수 있겠죠! 

그래서 보통 1번 객체는 제네릭을 활용하는 편이에요. (`T: Decodable`) 그리고 이를 사용하는 2번객체에선 원하는 모델(Network)로 매핑을 해서 다시 원하는 타입으로 사용하는 곳에 돌려주는 편이에요!

Baek-Gom-95 committed:
refactor : 네이밍 수정

- UrlCollection -> URLCollection
- JsonParser -> JSONParser
- Network -> ProductListResponse
- Page -> Product

bradheo65 committed:
rafactor: 초기화 구문 변경 및 불필요한 구문 제거

초기화 구문 변경
- URLRequest.init(url: url) ->  URLRequest(url: url)
불필요한 구문 제거
- print(json)

Baek-Gom-95 committed:
refactor : URLSession 처리로직 변경

Baek-Gom-95 committed:
refactor : URLSession Unit Test 리팩토링
